### ISSUE TITLE
Replace string references used in non user facing components in Feast

### ISSUE THREAD

**Issue by mrzzy** [feast-dev/feast#674]
## Background

Currently Feast uses string references in 3 areas:
- Feature Set References
- Feature References
- Subscriptions

As a result many implementations of string reference parsing and formatting has cropped up all over Feast's code:
- Feast Serving's [`RefUtil`](https://github.com/gojek/feast/blob/e4f8fe905d3c182416e1b69bf60bfbc0e8348f8a/serving/src/main/java/feast/serving/util/RefUtil.java)
- Feast Core's [`Store`](https://github.com/gojek/feast/blob/7a0ff91839757233d3af957b84096f9e2836386b/core/src/main/java/feast/core/model/Store.java#L127)
- Feast Ingestion's [`SpecUtil`](https://github.com/gojek/feast/blob/7a0ff91839757233d3af957b84096f9e2836386b/ingestion/src/main/java/feast/ingestion/utils/SpecUtil.java#L34)
- Feast Storage Writers: [`Redis`](https://github.com/gojek/feast/blob/7a0ff91839757233d3af957b84096f9e2836386b/storage/connectors/redis/src/main/java/feast/storage/connectors/redis/writer/RedisCustomIO.java#L185) [`RedisCluster`](https://github.com/gojek/feast/blob/7a0ff91839757233d3af957b84096f9e2836386b/storage/connectors/rediscluster/src/main/java/feast/storage/connectors/rediscluster/writer/RedisClusterCustomIO.java#L187).
- Feast Storage Retrievers [`Redis`](https://github.com/gojek/feast/blob/7a0ff91839757233d3af957b84096f9e2836386b/storage/connectors/redis/src/main/java/feast/storage/connectors/redis/retriever/RedisOnlineRetriever.java#L214) [`RedisCluster`](https://github.com/gojek/feast/blob/7a0ff91839757233d3af957b84096f9e2836386b/storage/connectors/rediscluster/src/main/java/feast/storage/connectors/rediscluster/retriever/RedisClusterOnlineRetriever.java#L218)
- etc...

## Problem

Changes to the feature/feature set references/subscriptions are difficult as we have to track down all the different string parsing/formatting implementations. Changes to these areas should not require a comb over the entire codebase to rectify all these parsing and formatting implementations.

Examples  of currently WIP and potential changes that is currently affected by this and why the problem is relevant:
- Feature Set version removal #386 
- Changes to the Feast concept & hierarchy and data model #479 
- Removal of Projects from Feature References when in the string form #631 

### Scope 

String references that should still be kept around:
- those exposed by the user facing API (ie the SDKs, config files) as they need to be humanly readable.
- those used by interface with components outside of feast (ie PostgreSQL, BigQuery, etc.), although non string references should be preferred if possible. 

## Possible Solution
Replace string references in non user facing components with their Protobuf equivalents.  
Examples of this include:
- Using message types directly in protobuf definitions:
```protobuf
message FeatureRow {
    repeated Field fields = 2;
    google.protobuf.Timestamp event_timestamp = 3;
    FeatureSetReference feature_set_ref = 8;
    string ingestion_id = 7;
}
```
>

- Using references as keys in maps directly.
```java
HashMap<FeatureSetReference, FeatureSetSpec> featureSetMap = new HashMap<>();
```

> :warning: There might be possible [caveats to this](https://stackoverflow.com/questions/34729968/protobuf-objects-as-keys-in-maps)
- Serialize references without having to extract its internals:
```java
String refStr = Base64.getEncoder().encodeToString(featureSetReference.toByteArray());
```

**Comment by woop** [comment]
Some comments:

 I agree on the Proto message type, but FeatureReferences should be a non-Proto model class ideally, with toProto and fromProto. This would allow us to also have a single `toStringRef()` or `fromStringRef()`. 

FeatureSetRef and FeatureRef seem very similar. I think we could possibly have the following after version removal
```
// Defines a composite key that refers to a unique FeatureSet
message FeatureSetReference {
    // Name of the project
    string project = 1;
    // Name of the FeatureSet
    string name = 2;
}
```

```
message FeatureReference {
    // Name of the feature set
    FeatureSetReference feature_set_ref = 1;

    // Name of the feature
    string name = 2;
}
```

Although this would increase nesting, which is also frowned upon by some.

**Comment by ches** [comment]
I mostly agree with the problem, but I'm not convinced about the solution. Swapping strings with encoded protobuf is merely trading one compatibility issue for another (less readable) one, in my mind. I think we're missing clear stated goals to answer whether the solution will meet them.

Definitely agree with @woop in general on avoiding the generated classes in domain logic if this is part of a solution—we've been moving away from that and although there is overhead (CPU + allocation/GC, and code maintenance), I believe it is a net win. The noted hash key behavior is one of many example pitfalls.

There are perhaps two related problems conflated, it could help to distinguish goals for solving them:

1. Code maintenance effort and fragility. This is primarily what the issue description elaborates so far.
1. Representations in persistence/serialization create challenges for evolving the data model.

For the first, I believe some goals for improving could be met sheerly through code factoring, encapsulation of the domain concepts of references as first-class types instead of `String`s pervading.

The second one is harder, I'm not sure there is an infinitely flexible solution with reasonable complexity (though I'd be glad to hear suggestions), rather some design decisions need to be reached on #479 that we're ready to live with for awhile and/or have a plan for evolving. It's a conceptual problem before it's a technical one. If we're ready to go with https://github.com/gojek/feast/issues/479#issuecomment-615544705 as the first steps, I think that should be a clear part of specs in this issue or an explicit non-goal, since #631 seems meant to cover that at least in part.

With this perspective, the first problem could be appreciably addressed without being blocked by exhaustive answers for #479, except that it might overlap a lot with work for #631. Once goals of this issue are clearly established, it might just become a subtask of #631. 

Definitively solving the second is blocked by #479 in such ways that it could put this issue at risk of not being closeable in the near term.

**Comment by ches** [comment]
> For the first, I believe some goals for improving could be met sheerly through code factoring, encapsulation of the domain concepts of references as first-class types instead of `String`s pervading.

Putting this another way, I'd like to explore how this might look with only code changes that are not proto schema changes, _then_ evaluate if and when proto changes should be made. If there can be some win from the first step, it seems more in line with @woop's inclination in https://github.com/gojek/feast/issues/479#issuecomment-615544705 to be conservative about the conceptual + data model direction, to defer until we learn more.
